### DISCUSSION TITLE
Можно сгенерировать такое дерево?

### DISCUSSION THREAD

**Discussion by Makhmut2445** [totumonline/totum-ru-issues-and-discussions#293] [Как это сделать?] [answered]
![image](https://user-images.githubusercontent.com/103171127/208231307-ff0ff394-02cb-46e5-a1c2-d6baae6ad5a6.png)

Пример:

Договор М-33 (дерево)
   Проект ВМБД-13 (дерево)
      Состав ИД (дерево)
         Книга 1 (дерево)
         ТЧ Саидов (строка)
         ГЧ Филимонов (строка)

**Comment by ttm-alexeypolunin** [comment, marked as answer]
<img width="272" alt="Снимок экрана 2022-12-20 в 17 45 50" src="https://user-images.githubusercontent.com/55755565/208694077-00b1c890-9ecd-469a-a1c2-4b98fb12ca30.png">

Можно — код немножко дикий, но можно.

В поле Селект-дерево сейчас неправильно обрабатываются пустые строки — починим я напишу как это сделать.

**Reply by ttm-alexeypolunin** [reply]
**Работает начиная с 4.8.51**

Идея в следующем:

Код дерева (как и Селекта) можно хранить просто как структуру в поле. Нам это потребуется тк такое конское дерево считать при каждом открытии таблицы затруднительно (с ростом его размера оно будет считаться дольше и дольше).

Поэтому мы делаем где-то поле тип данные и в нем делаем код, который собирает дерево. Пересчитываем через какой-то интервал по крону.

```
= : rowKeysRemove(row: $train_final; key: "id"; recursive: true)

train_final: listtrain(list: $replace)

replace: listReplace(list: $add_parent; action: $val; value: "value")
	val: listAdd(list: $innerreplace; item: $#value)
				innerreplace: rowKeysReplace(row: $train; from: json`["book","base","nazvanie"]`; to: json`["title","value","title"]`; recursive: true)

		train: listTrain(list: $inner)
			inner: listReplace(list: $zayavki; action: $add_z; value: "value2")
					zayavki: selectList(table: 'zayavki'; field: 'id'; where: 'dogovor' = $#value[id]; order: 'nazvanie' asc)
				
				add_z: listAdd(item: $zayavka; list: $third_replace_train)
					zayavka: rowAdd(row: $sel_z; field: "parent" = $#value[value])
						sel_z: var(name: "sel_z"; value: $sel_sel)
							sel_sel: selectRow(table: 'zayavki'; field: 'nazvanie'; field: 'base'; where: 'id' = $#value2)
					
					third_replace_train: listTrain(list: $third_replace)
						third_replace: listReplace(list: $list_of_tables; action: $add_t; value: "value3")
								list_of_tables: rowListCreate(field: "base" = $listnames; field: "title" = $listtitles)
									listnames: listCreate(item: str`"tip_1-" + $#sel_z[base]`; item: str`"tip_2-" + $#sel_z[base]`)
									listtitles: json`["Тип 1","Тип 2"]`
					
					add_t: listAdd(item: $value3_parent; list: $val2)
						value3_parent: rowAdd(row: $#value3; field: "parent" = $#sel_z[base])
						val2: rowListAdd(rowlist: $type; field: "parent" = $#value3[base])
							type: selectRowList(table: $value3_base[0]; cycle: $#value2; field: 'book'; field: 'base'; order: 'book' asc)
								value3_base: strSplit(str: $#value3[base]; separator: "-")
				
add_parent: rowListAdd(rowlist: $rename; field: "parent" = "")
	rename: rowKeysReplace(row: $first_list; from: json`["dogovor","base"]`; to: json`["title","value"]`; recursive: true)
		first_list: selectRowList(table: 'dogovora'; field: 'base'; field: 'dogovor'; field: 'id'; order: 'dogovor' asc)

```

> Код этот может быть и поаккуратней, но у меня времени не было его причесывать

### Таблицы:

**Договора:**

<img width="612" alt="Снимок экрана 2022-12-27 в 17 41 13" src="https://user-images.githubusercontent.com/55755565/209682238-c755eae3-ebe9-475d-87f1-eb5c11de572b.png">

**Заявки:**

<img width="639" alt="Снимок экрана 2022-12-27 в 17 42 28" src="https://user-images.githubusercontent.com/55755565/209682351-5258b2a3-ad98-45a6-8eff-071f39733901.png">

**Таблицы в цикле:**

<img width="386" alt="Снимок экрана 2022-12-27 в 17 43 19" src="https://user-images.githubusercontent.com/55755565/209682439-51b1f62b-c261-48ca-ae8b-051bc583ea96.png">

**Обратите внимание, нам везде потребовались уникальные основания на основе id!**

Как результат у нас вот такая структура:

```
[
  {
    "title": "Книга 1",
    "value": "z-1-t-1-b-1",
    "parent": "tip_1-z-1"
  },
  {
    "title": "Книга 2",
    "value": "z-1-t-1-b-2",
    "parent": "tip_1-z-1"
  },
  {
    "title": "Тип 1",
    "value": "tip_1-z-1",
    "parent": "z-1"
  },
  {
    "title": "Книга 1",
    "value": "z-1-t-2-b-1",
    "parent": "tip_2-z-1"
  },
  {
    "title": "Книга 2",
    "value": "z-1-t-2-b-2",
    "parent": "tip_2-z-1"
  },
  {
    "title": "Тип 2",
    "value": "tip_2-z-1",
    "parent": "z-1"
  },
  {
    "title": "Заявка 1",
    "value": "z-1",
    "parent": "d-1"
  },
  {
    "title": "Книга 1",
    "value": "z-3-t-1-b-1",
    "parent": "tip_1-z-3"
  },
  {
    "title": "Книга 2",
    "value": "z-3-t-1-b-2",
    "parent": "tip_1-z-3"
  },
  {
    "title": "Тип 1",
    "value": "tip_1-z-3",
    "parent": "z-3"
  },
  {
    "title": "Книга 1",
    "value": "z-3-t-2-b-1",
    "parent": "tip_2-z-3"
  },
  {
    "title": "Книга 2",
    "value": "z-3-t-2-b-2",
    "parent": "tip_2-z-3"
  },
  {
    "title": "Тип 2",
    "value": "tip_2-z-3",
    "parent": "z-3"
  },
  {
    "title": "Заявка 3",
    "value": "z-3",
    "parent": "d-1"
  },
  {
    "title": "Договор 1",
    "value": "d-1",
    "parent": ""
  },
  {
    "title": "Книга 1",
    "value": "z-2-t-1-b-1",
    "parent": "tip_1-z-2"
  },
  {
    "title": "Книга 2",
    "value": "z-2-t-1-b-2",
    "parent": "tip_1-z-2"
  },
  {
    "title": "Тип 1",
    "value": "tip_1-z-2",
    "parent": "z-2"
  },
  {
    "title": "Книга 1",
    "value": "z-2-t-2-b-1",
    "parent": "tip_2-z-2"
  },
  {
    "title": "Книга 2",
    "value": "z-2-t-2-b-2",
    "parent": "tip_2-z-2"
  },
  {
    "title": "Тип 2",
    "value": "tip_2-z-2",
    "parent": "z-2"
  },
  {
    "title": "Заявка 2",
    "value": "z-2",
    "parent": "d-2"
  },
  {
    "title": "Договор 2",
    "value": "d-2",
    "parent": ""
  }
]

```

И мы ее подтаскиваем в Код селекта просто беря через select:

<img width="649" alt="Снимок экрана 2022-12-27 в 17 46 36" src="https://user-images.githubusercontent.com/55755565/209683005-197a2c2d-141c-4381-a49d-9f18e704552a.png">

Последняя колонка в задаче — это будут строки в таблице в которой используется это дерево. А в дере у них основание соответствующей книги. У книг уникальные номера зависящие от цикла, таблицы и id строки — `z-2-t-2-b-2`
